src/test/shell/bazel/target_compatible_with_test.sh

```diff
@@ -0,0 +1,696 @@
+# starting with "msys", and "*" matches everything (it's the default case).
+case "$(uname -s | tr [:upper:] [:lower:])" in
+msys*)
+  # As of 2019-01-15, Bazel on Windows only supports MSYS Bash.
+  declare -r is_windows=true
+  ;;
+*)
+  declare -r is_windows=false
+  ;;
+esac
+
+if "$is_windows"; then
+  export MSYS_NO_PATHCONV=1
+  export MSYS2_ARG_CONV_EXCL="*"
+fi
+
+function set_up() {
+  mkdir -p target_skipping || fail "couldn't create directory"
+  cat <<EOT > target_skipping/pass.sh || fail "couldn't create pass.sh"
+#!/bin/bash
+exit 0
+EOT
+  chmod +x target_skipping/pass.sh
+
+  cat <<EOT > target_skipping/fail.sh || fail "couldn't create fail.sh"
+#!/bin/bash
+exit 1
+EOT
+  chmod +x target_skipping/fail.sh
+
+  cat <<EOT > target_skipping/BUILD || fail "couldn't create BUILD file"
+# We're not validating visibility here. Let everything access these targets.
+package(default_visibility = ["//visibility:public"])
+
+constraint_setting(name = "not_compatible_setting")
+
+constraint_value(
+    name = "not_compatible",
+    constraint_setting = ":not_compatible_setting",
+)
+
+constraint_setting(name = "foo_version")
+
+constraint_value(
+    name = "foo1",
+    constraint_setting = ":foo_version",
+)
+
+constraint_value(
+    name = "foo2",
+    constraint_setting = ":foo_version",
+)
+
+constraint_value(
+    name = "foo3",
+    constraint_setting = ":foo_version",
+)
+
+constraint_setting(name = "bar_version")
+
+constraint_value(
+    name = "bar1",
+    constraint_setting = "bar_version",
+)
+
+constraint_value(
+    name = "bar2",
+    constraint_setting = "bar_version",
+)
+
+platform(
+    name = "foo1_bar1_platform",
+    parents = ["@local_config_platform//:host"],
+    constraint_values = [
+        ":foo1",
+        ":bar1",
+    ],
+)
+
+platform(
+    name = "foo2_bar1_platform",
+    parents = ["@local_config_platform//:host"],
+    constraint_values = [
+        ":foo2",
+        ":bar1",
+    ],
+)
+
+platform(
+    name = "foo1_bar2_platform",
+    parents = ["@local_config_platform//:host"],
+    constraint_values = [
+        ":foo1",
+        ":bar2",
+    ],
+)
+
+platform(
+    name = "foo3_platform",
+    parents = ["@local_config_platform//:host"],
+    constraint_values = [
+        ":foo3",
+    ],
+)
+
+sh_test(
+    name = "pass_on_foo1",
+    srcs = ["pass.sh"],
+    target_compatible_with = [":foo1"],
+)
+
+sh_test(
+    name = "fail_on_foo2",
+    srcs = ["fail.sh"],
+    target_compatible_with = [":foo2"],
+)
+
+sh_test(
+    name = "pass_on_foo1_bar2",
+    srcs = ["pass.sh"],
+    target_compatible_with = [
+        ":foo1",
+        ":bar2",
+    ],
+)
+
+sh_binary(
+    name = "some_foo3_target",
+    srcs = ["pass.sh"],
+    target_compatible_with = [
+        ":foo3",
+    ],
+)
+EOT
+
+  cat <<EOT > target_skipping/WORKSPACE || fail "couldn't create WORKSPACE"
+EOT
+}
+
+function tear_down() {
+  bazel shutdown || :
```

Comment:
Is this needed for the tests to pass?

Reply:
Nope, removed.

Reply:
I'd put that there because on Windows I kept getting errors like `rm: target_skipping: Device or resource busy.`.
They went away when I put the shutdown there.

Should have left a comment there, sorry. Either way, the tests on Windows still pass, but they're just noisy without this line.

Reply:
Would you mind filing a separate issue about the noise? We should fix it for all tests at once.

Reply:
@katre, please see #12397.